File: problems/1004/1004.max-consecutive-ones-iii.go
/*
 * @lc app=leetcode id=1004 lang=golang
 *
 * [1004] Max Consecutive Ones III
 */

package solution

// @lc code=start
func longestOnes(nums []int, k int) int {
	l, r, curr, res := 0, 0, 0, 0

	for i := 0; i < len(nums); i++ {
		if nums[i] == 0 {
			curr++
		}

		for curr > k {
			if nums[l] == 0 {
				curr--
			}
			l++
		}

		r++
		res = max(res, r-l)
	}

	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}

	return b
}

// @lc code=end

File: problems/1004/1004.max-consecutive-ones-iii.java
/*
 * @lc app=leetcode id=1004 lang=java
 *
 * [1004] Max Consecutive Ones III
 */

// @lc code=start
class Solution {
    public int longestOnes(int[] nums, int k) {
        /*
         * max number of consecutive 1s if can flip at most k 0s
         *
         * intuition:
         * - sliding window
         * - such that nums[l, r] has k flips
         *
         * O(N) time, O(1) space
         */

        int ans = 0;
        int curr = 0;
        int l = 0;
        int r;

        for (r = 0; r < nums.length; r++) {
            if (nums[r] == 1) {
                continue;
            }

            if (curr + 1 <= k) {
                curr++;
                continue;
            }

            if (r - l > ans)
                ans = r - l;

            while (nums[l] == 1 && l <= r) {
                l++;
            }
            l++;
        }

        if (r - l > ans) {
            ans = r - l;
        }

        return ans;
    }
}
// @lc code=end
